### ISSUE TITLE
Added hashCode and == operator to all params

### ISSUE THREAD

**Issue by stevenroose** [PointyCastle/pointycastle#82]
_From @stevenroose on March 29, 2014 15:14_

I needed this for ScryptParams and decided to add it to all of them.

Only one still has an issue (I placed a TODO): ParametersWithRandom. For this param, I only took the runtimeType of the SecureRandom into account. Otherwise, I'd have to implement hashCodes and equals for all randoms as well, which would cascade in adding them to all block ciphers.

I know runtimeType is not enough. For now, I clearly documented this drawback so I hope it's ok.

I had to add dependency for package:collection to pubspec.

_Copied from original issue: izaera/cipher/pull/82_

**Comment by stevenroose** [comment]
_From @izaera on March 29, 2014 15:21_

:-o  Out of curiosity, what do you need that for?

I'll merge it, thanks. Regarding the ParametersWithRandom I think it makes sense to just ignore the SecureRandom (even not doing the runtimeType check). A SecureRandom is not data, but an algorithm, so equals for it doesn't make too much sense, I think...

**Comment by stevenroose** [comment]
When a private key is encrypted, the KeyCrypter used to encrypt and decrypt
the key is stored along with the encrypted bytes. This means that the
KeyCrypter is just as much a part of the actual key as the encrypted bytes.

For that, to support checking of two keys are equal, the KeyCrypter should
be comparable as well. The currently only existing KeyCrypter is a Scrypt
one, which is completely defined by the ScryptParams it contains.

So basically, two keys are the same if the encrypted bytes and the IV are
the same AND if the scrypt params are the same.

**Comment by stevenroose** [comment]
_From @izaera on March 29, 2014 15:35_

I see.

I guess that could be better done when ASN.1 encoding is available. You can encrypt the parameters and type of cipher used along with the output (as RSA signing does, for example). Then it's a matter of comparing two bundles of bit data.

As an aside, in that context it would make more sense not to check the SecureRandom in ParametersWithRandom as it wouldn't be serialized to ASN.1.

**Comment by stevenroose** [comment]
(Is it possible to make a change to a PR in GitHub?)

If no, you can remove the runtimeType check from PWR then :)

**Comment by stevenroose** [comment]
_From @izaera on March 29, 2014 15:47_

You can keep pushing to PRs. In the end they are just branches in your repo + a fancy message telling the destination user to pull from it.

But never mind, I'll do it while I merge it. :-)
